airbyte-ci/connectors/pipelines/pipelines/helpers/execution/run_steps.py

```diff
@@ -27,16 +27,78 @@ class InvalidStepConfiguration(Exception):
     pass
 
 
+def _get_dependency_graph(steps: STEP_TREE) -> Dict[str, List[str]]:
+    """
+    Get the dependency graph of a step tree.
+    """
+    dependency_graph: Dict[str, List[str]] = {}
+    for step in steps:
+        if isinstance(step, StepToRun):
+            dependency_graph[step.id] = step.depends_on
+        elif isinstance(step, list):
+            nested_dependency_graph = _get_dependency_graph(list(step))
+            dependency_graph = {**dependency_graph, **nested_dependency_graph}
+        else:
+            raise Exception(f"Unexpected step type: {type(step)}")
+
+    return dependency_graph
+
+
+def _get_transitive_dependencies_for_step_id(
+    dependency_graph: Dict[str, List[str]], step_id: str, visited: Optional[Set[str]] = None
+) -> List[str]:
+    """Get the transitive dependencies for a step id.
+
+    Args:
+        dependency_graph (Dict[str, str]): The dependency graph to use.
+        step_id (str): The step id to get the transitive dependencies for.
+        visited (Optional[Set[str]], optional): The set of visited step ids. Defaults to None.
+
+    Returns:
+        List[str]: List of transitive dependencies as step ids.
+    """
+    if visited is None:
+        visited = set()
+
+    if step_id not in visited:
+        visited.add(step_id)
+
+        dependencies: List[str] = dependency_graph.get(step_id, [])
+        for dependency in dependencies:
+            dependencies.extend(_get_transitive_dependencies_for_step_id(dependency_graph, dependency, visited))
+
+        return dependencies
+    else:
+        return []
+
+
 @dataclass
 class RunStepOptions:
     """Options for the run_step function."""
 
     fail_fast: bool = True
     skip_steps: List[str] = field(default_factory=list)
+    keep_steps: List[str] = field(default_factory=list)
     log_step_tree: bool = True
     concurrency: int = 10
     step_params: Dict[CONNECTOR_TEST_STEP_ID, STEP_PARAMS] = field(default_factory=dict)
 
+    def __post_init__(self) -> None:
+        if self.skip_steps and self.keep_steps:
+            raise ValueError("Cannot use both skip_steps and keep_steps at the same time")
+
+    def get_step_ids_to_skip(self, runnables: STEP_TREE) -> List[str]:
```

Comment:
❗This is all great. Only thing thats missing is some tests to ensure weve got the traversal correct